runtime/doc/lua.txt

```diff
@@ -1714,11 +1714,43 @@ vim.region({bufnr}, {pos1}, {pos2}, {regtype}, {inclusive})
         `endcol` is exclusive, and whole lines are returned as
         `{startcol,endcol} = {0,-1}`.
 
-vim.schedule_wrap({cb})                                  *vim.schedule_wrap()*
-    Defers callback `cb` until the Nvim API is safe to call.
+vim.schedule_wrap({callback})                            *vim.schedule_wrap()*
+    Returns a function which calls {callback} via |vim.schedule()|.
+
+    Useful to "wrap" existing functions, ensuring they are only called when
+    the Nvim API is safe to call. The returned function passes all arguments
+    to {callback}.
+
+    Example: >lua
+        ---
+        -- Automatically turning an unsafe callback into a safe callback.
+        -- The `event` argument given to autocmd's callback is passed along.
+        local function my_function(event)
+          -- Some "unsafe" code ...
+        end
+        vim.api.nvim_create_autocmd("TextChangedI", {
```

Comment:
autocmd is not the best example here as the API is already safe in an autocmd. Deferring autocmds can be useful, but _not_ for the reason stated in the comment just above, which makes this confusing. An example with `vim.uv` callbacks would be better right here.

Reply:
How about

```lua
function notify_readable(_err, readable)
  -- Without schedule_wrap, this would raise the error "nvim_echo must not be called in a lua loop callback"
  vim.notify("readable? " .. tostring(readable))
end
vim.uv.fs_access(vim.fn.stdpath("config"), "R", vim.schedule_wrap(notify_readable))
```

- fs_access is pretty understandable without knowing much about uv
- Passes an argument to callback
- Not much pre-work fanfare (eg, no other handles need to be created etc)

A timer is used in other examples, but that does not pass a value back to the callback which I think is helpful to include.